caddy/websocket_handler.go

```diff
@@ -0,0 +1,181 @@
+// Copyright 2024 The Outline Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package outlinecaddy
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"io"
+	"log/slog"
+	"net"
+	"net/http"
+	"time"
+
+	"github.com/Jigsaw-Code/outline-sdk/transport"
+	"github.com/caddyserver/caddy/v2"
+	"github.com/gorilla/websocket"
+	"github.com/mholt/caddy-l4/layer4"
+)
+
+const wsModuleName = "layer4.handlers.websocket"
+
+func init() {
+	caddy.RegisterModule(ModuleRegistration{
+		ID:  wsModuleName,
+		New: func() caddy.Module { return new(WebSocketHandler) },
+	})
+}
+
+type ConnectionType string
+
+const connectionTypeStream ConnectionType = "stream"
+const connectionTypePacket ConnectionType = "packet"
+
+// WebSocketHandler implements a Caddy plugin for WebSocket connections.
+type WebSocketHandler struct {
+	Type ConnectionType `json:"type,omitempty"`
+
+	logger *slog.Logger
+	u      websocket.Upgrader
+}
+
+var (
+	_ caddy.Provisioner  = (*WebSocketHandler)(nil)
+	_ caddy.Validator    = (*WebSocketHandler)(nil)
+	_ layer4.NextHandler = (*WebSocketHandler)(nil)
+)
+
+func (*WebSocketHandler) CaddyModule() caddy.ModuleInfo {
+	return caddy.ModuleInfo{ID: wsModuleName}
+}
+
+// Provision implements caddy.Provisioner.
+func (h *WebSocketHandler) Provision(ctx caddy.Context) error {
+	h.logger = ctx.Slogger()
+	return nil
+}
+
+func (h *WebSocketHandler) Validate() error {
+	if h.Type != connectionTypeStream && h.Type != connectionTypePacket {
+		return fmt.Errorf("unsupported connection type: %s", h.Type)
+	}
+	return nil
+}
+
+// Handle implements layer4.NextHandler.
+func (h *WebSocketHandler) Handle(cx *layer4.Connection, next layer4.Handler) error {
+	req, err := http.ReadRequest(bufio.NewReader(cx))
```

Comment:
You are reproducing HTTP logic here. You don't need this if you move it to the HTTP app.

Reply:
You'll need a HTTP handler like this: https://github.com/Jigsaw-Code/outline-sdk/blob/f521e2b520d1ea93d4987b159c8b217ce8ea9e65/x/examples/ws2endpoint/main.go#L71